smith.1
.TH SMITH 1 
.SH NAME
smith \- a Python\-based build system for font development

http://github.com/silnrsi/smith

.SH SYNOPSIS
.B smith
.RI [ options ]
.SH DESCRIPTION
This manual page documents the
.B smith
commands.

.B  For more details please refer to the complete smith manual.


.PP
\fBSmith\fP is a Python\-based framework for building, testing and maintaining fonts.
It is based on waf. A smith project contains a wscript file (a python script) with the various parameters at the root of the folder containing the font sources.
.SH TARGET OPTIONS
.SH The following targets are available
.TP
start \- populate your new project folder with templates.
.TP
configure \- configure your project and check all the needed dependencies are available
.TP
build  \- launch the build
.TP
pdfs  \- generate pdfs (proofs for review including test of the smart features)
.TP
clean \- delete generated files
.TP
distclean \- delete generated folders
.TP
zip  \- create a versioned release zip (with Windows line-endings)
.TP
tarball  \- create a versioned release tarball (with Linux line-endings)
.TP
release  \- create a versioned release zip and tarball
.TP
test  \- run the project\-specific test suite 
.TP
xtest  \- run the project\-specific rendering test suite 
.TP
alltests  \- run all the project\-specific tests 
.TP
fbchecks  \- run fontbakery QA tests via SIL profile 
.PP
fontspector  \- run fontspector QA tests via SIL profile 
.PP
differ  \- run diffenator2 QA by comparing built fonts against references (STANDARDS) 
.PP
graide \- generate a graide (GRAphite Integrated Development Environment) project file 
.TP
ots \- run the OpenType Sanitiser 
.PP
version \- find out which version of smith your have 
.PP
checksums \- generate checksums (SHA512SUMS.txt) for available artefacts in the release folder
.PP
sign \- generate .asc detached signatures for available artefacts in the release folder (needs a GPG key)
.PP
.SH The following options are available
.PP
-vvv increase verbosity
-r --release change to release mode
.PP
For details of the available options see the complete smith manual.


.SH AUTHOR
.PP
smith was written by Martin Hosken <martin_hosken@sil.org>.
.TP
with contributions from Nicolas Spalinger <nicolas_spalinger@sil.org> and others.
.TP
SIL WSTech \- http://software.sil.org/wstech
SIL Global \- http://www.sil.org 
.PP


ChangeLog
NEW IN WAF 1.6.8
----------------
* Found a typo that prevented Waf from running on Pypy-trunk
* Fixed a typo in Utils.py affecting Win32 platforms (copystat) #1029
* Fixed a minor bug in the Eclipse project generator
* Make the xlc/xlc++ compiler detection more accurate by looking at the version number (#1022)

NEW IN WAF 1.6.7
----------------

* Provide more diagnostic for invalid build groups #914
* Various enhancements to msvs.py
* Read MSVC_VERSIONS and MSVC_TARGETS from the command-line
* Minor cross-compiler detection fix on msvc.py
* Fix the redirections with pipes (waf configure > log)
* Do not display runnable_status exceptions when running with -k
* Let -k stop at the first runnable_status error and -kk run even further
* Merge the add_object extension in the main line (source='file.o')
* Make update_outputs more robust with changes in the task definition #1017
* Qt4 detection on Win32

NEW IN WAF 1.6.6
----------------

* Fix the performance regression related to #974

NEW IN WAF 1.6.5
----------------

* More documentation
* Re-enable the colors for msys
* Add the .ui files for the qt4 translations
* Fix the conf.check_large_file() test
* Fix the conf.check_library() in C++ mode #955
* Improve the latex scanner to avoid depending on generated files #943
* Remove the @file processing from the winrc tasks
* Fix the python detection using python-config
* Add the missing default includes and defines to the moc command
* Improve support for hierarchical go-lang packages #953
* Fix the gfortran verbose flag detection on Windows
* Fix the support of fortran import libraries #950
* Added a fix for running with Python 2.4 on Windows #949
* Limited support for IronPython
* Support for older Visual Studio versions (VC6) #952
* New waf.bat file #964
* New method ConfigSet.keys
* New Visual Studio and Eclipse CTD project generators (waflib/extras)
* New lru_cache tool for use with WAFCACHE (waflib/extras)

NEW IN WAF 1.6.4
----------------

* Fix the Python detection on win32 #883
* Optimize the Python file installation #892
* Force +x permissions on installed fortran programs #893
* Qt library detection fixes #895
* Ensure that unit tests are executed only after the symlinks are created
* Fix the precedence constraints for classes ending in _task #896
* Support the install_path attribute with add_pcfile #913
* Make the goprograms executable when installed #928
* Allow non-python data files in the waf executable #927
* Enforce a build order based on the scanner results #777, #922
* Multiple msvc detection fixes #907 #910 #923 #924 #926
* Fix the -of flag append with dmd #917
* Boost detection fixes #920
* Support newer javac compilers #921
* Fix the execution on python 2.3 for: "waf update", msvc.py, fc.py
* Improve the support for mac applications (demos/mac_app)
* Better default regexps in "waf step"
* New error check for tasks creating the same nodes or having the same identifiers (waf -v)
* New variables conf.env.NO_LOCK_IN_TOP/OUT/RUN for special projects (top='..')
* New example on avoiding rebuilds when moving a project (playground/relocate)
* Improve go-lang support for cgo-packages (fixes #932)
* Fix the progress bar on cmd and msys

NEW IN WAF 1.6.3
----------------

* Fixed the interaction of Fortran configuration tests and WAFCACHE #842
* Various documentation fixes
* Set the PYTHONARCHDIR variable for installing python extensions
* Fixed the Waf file creation with --zip-type=gz (bunzip2 was not replaced by gzip -d)
* Fixed multiple problems in the call to TaskGen.declare_chain(...) #850
* Fixed the task attribute 'vars' which might cause unnecessary rebuilds #852
* Return the value of post_check(...) in conf.check(...) #857
* Rewrite the boost tool (waflib/extras/boost.py) #814, #454, #424
* More fortran file extensions: .for, .FOR #867
* Searching above the root nodes no longer raise exceptions #868
* Msvc detection fixes for non-utf8 encodings #873
* Fixed the destdir processing on Windows #874
* Stop changing the flags on waf -v (make the errors more visible) #875
* Fixed the resource file compilation on Windows #876
* Fixed the vala file installation #881
* New system of plugins for C/C++/Fortran compilers (files named c_* in waflib/extras/)
* New examples of interaction between Waf and existing makefiles (playground/)
* New names for @before/@after: @before_method/@after_method

NEW IN WAF 1.6.2
----------------

* Support for C# debugging files #797
* Add -relocation-model=pic for shared libraries on ldc
* Fixed 'waf dist' for tar files on python 3 #799
* Make the latex scanner recursive #798
* Enable the packing of non-python files in the waf file #802
* Improve the feature sniffing for hybrid programs/libraries #800
* New apidocs + tutorial in Sphinx format
* Add the CPPFLAGS from os.environ #803
* Create the java output directory anywhere #806
* Enable the .luac file installation
* Process Qt translation files
* Detect when the folders were copied and prompt for "waf configure"
* Parse_flags for the *-config outputs on windows (backslashes) #811
* Fix the doxygen task build order + improve the dependency scanner #821
* Various msvc fixes #819, #826, #825
* Ported the batch executor to waf 1.6 (batched_cc)
* New tools: erlang, scala
* Moved conf.multicheck(..) from playground to the library
* New parameter to avoid reading the same scripts: bld.recurse(dir, once=True)
* Detect invalid method calls in 'waf -v' such as env.append/env.add/env.prepend
* New manifest option for jar targets #832

NEW IN WAF 1.6.1
----------------

* Fixed the method check_waf_version  #764
* Fixed the name in ctx.recurse(name) #769
* Stop caching the install tasks and tasks that have no outputs #770
* Fix the log in Context.cmd_and_log() when calling with "quiet" #778
* c_preproc exception when a file has the same name as a directory #777
* 'intltool_po' does not install the .mo files #782
* 'intltool_in' was broken #792

src/smithlib/smith.py
#!/usr/bin/env python3
''' smith module '''
__url__ = 'http://github.com/silnrsi/smith'
__copyright__ = 'Copyright (c) 2011-2025 SIL Global (http://www.sil.org)'
__author__ = 'Alan Ward, Martin Hosken, Nicolas Spalinger'
__license__ = 'Released under the 3-Clause BSD License (http://opensource.org/licenses/BSD-3-Clause)'


from smithlib import smith
from waflib import Scripting, Context, Task
from .wafplus import *
import sys, os, shlex, re
from optparse import OptionParser

waflib_dir = os.path.dirname(os.path.abspath(Scripting.__file__)).split(os.sep)
waf_dir = os.sep.join(waflib_dir[:-1])

def main():
    Scripting.waf_entry_point(os.getcwd(), Context.WAFVERSION, waf_dir)


def formatvars(s, kw=None):
    if isinstance(s, deferred_class):
        return s(kw)
    elif not s:
        return s

    def cvt(m):
        return kw.get(m.group(1), '${' + m.group(1) + '}')
    if isinstance(s, str):
        if kw is None:
            return s
        else:
            return re.sub(r'\${([a-zA-Z:_/]*)}', cvt, s)
    if hasattr(s, 'items'):
        return dict((k, formatvars(v, kw)) for k,v in s.items())
    elif hasattr(s, '__len__'):
        return [formatvars(x, kw) for x in s]
    return s

class deferred_class(object):
    def __init__(self, c, a, kw):
        self.c = c
        self.a = a
        self.kw = kw

    def __call__(self, kw=None):
        a = [formatvars(x, kw) for x in self.a]
        kwr = dict((k, formatvars(v, kw)) for k,v in self.kw.items())
        return self.c(*a, **kwr)

def defer(c):
    def g(*a, **kw):
        return deferred_class(c, a, kw)
    return g

def undeffered(c):
    def undeffer(*a, **kw):
        newa = list(map(formatvars, a))
        newkw = dict((k, formatvars(v)) for k,v in kw.items())
        return c(*newa, **newkw)
    return undeffer

def isdeferred(c):
    return isinstance(c, deferred_class)

def initval(v):
    return v() if type(v) == 'deferred_class' else v

def initobj(self, kw):
    for k, v in kw.items():
        setattr(self, k, initval(v))

class _create(str) :

    isGenerated = 1
    def __new__(self, tgt, *cmds, **kw) :
        if hasattr(tgt, 'len') : tgt = tgt[0]
        return str.__new__(self, initval(tgt))

    def __init__(self, tgt, *cmds, **kw) :
        tgt = initval(tgt)
        self.cmds = list(map(initval, cmds))
        if len(cmds) > 0 :
            res = cmds[0]()(tgt) if isdeferred(cmds[0]) else cmds[0](tgt)
            res[2].update(kw)
            rule(res[0], res[1], tgt, **res[2])
        for c in cmds[1:] :
            res = c()(tgt) if isdeferred(c) else c(tgt)
            res[2].update(kw)
            modify(res[0], tgt, res[1], **res[2])

    def get_sources(self, ctx) :
        res = []
        for c in self.cmds :
            res.extend(c().get_sources(ctx) if isdeferred(c) else c.get_sources(ctx))
        return res

create = defer(_create)

class _process(_create) :

    def __new__(self, tgt, *cmds, **kw) :
        if not hasattr(tgt, 'len') and os.path.exists(tgt) :
            return super(_process, self).__new__(self, os.path.join("tmp", os.path.basename(tgt)), *cmds, **kw)
        else : return super(_process, self).__new__(self, tgt, *cmds, **kw)

    def __init__(self, tgt, *cmds, **kw) :
        tgt = initval(tgt)
        # if tgt exists in source tree, then become create(munge(tgt), cmd("${CP} ${SRC} ${TGT}", [tgt]), *cmds, **kw)
        if os.path.exists(tgt) :
            cmds = [cmd("cp -r ${SRC} ${TGT}", [tgt])] + list(cmds)
            tgt = os.path.join("tmp", os.path.basename(tgt))
            super(_process, self).__init__(tgt, *cmds, **kw)
        else :
            self.cmds = list(map(initval, cmds))
            for c in cmds :
                res = c(tgt)
                if 'late' not in res[2]:
                    res[2]['late'] = 1
                elif not res[2]['late']:
                    del res[2]['late']
                res[2].update(kw)
                modify(res[0], tgt, res[1], **res[2])

process = defer(_process)

class test(_process) :

    def __init__(self, tgt, *cmds, **kw) :
        kw['nochange'] = 1
        super(test, self).__init__(tgt, *cmds, **kw)


class _cmd(object) :
    def __init__(self, c, inputs = [], **kw) :
        self.c = initval(c)
        if not isinstance(inputs, list):
            inputs = [inputs]
        self.inputs = list(map(initval, inputs))
        self.opts = dict((k, initval(v)) for k, v in kw.items())

    def __call__(self, tgt) :
        return (self.c, self.inputs, self.opts)

    def get_sources(self, ctx) :
        res = get_all_sources(self, ctx, 'inputs')
        c = shlex.split(self.parse(ctx))[0]
        if not os.path.isabs(c) :
            n = ctx.bldnode.find_node(c)
            if n and not n.is_child_of(ctx.bldnode) :
                pat = n.abspath()
                res.append(n.srcpath())
        return res

    def parse(self, ctx, kw = None) :
        if kw is None : kw = self.opts
        def repl(match) :
            g = match.group
            if g('dollar') : return '$'
            elif g('backslash') : return '\\\\'
            elif g('subst') :
                if not g('code') and g('var') in kw :
                    return kw[g('var')]
                else : return '${' + g('var') + g('code') + '}'
        return Task.reg_act.sub(repl, self.c)

    def build(self, ctx, inputs, tgt, **kw) :
        return ctx(rule = self.parse(ctx, kw), source = inputs, target = tgt, **kw)

cmd = defer(_cmd)

def isList(l) :
    return (isinstance(l, (list, tuple)) or
            not hasattr(l, 'strip') and
            not hasattr(l, 'keys') and
                hasattr(l, '__getitem__') and
                hasattr(l, '__iter__'))

def _get_source(self, ctx, a) :
    n = ctx.path.find_node(a)
    if not hasattr(ctx, 'bldnode') : # then we are a simple context
        if not n :
            return []
        elif os.path.isdir(n.abspath()) :
            return [x.path_from(ctx.path) for x in n.find_nodes()]
        else :
            return [n.path_from(ctx.path)]
    elif not n :
        return []
    elif os.path.isdir(n.abspath()) :
        return [x.srcpath() for x in n.find_nodes()]
    else :
        return [n.srcpath()]

def get_all_sources(self, ctx, *attrs) :
    res = []
    for a in (getattr(self, x, None) for x in attrs) :
        if a :
            if hasattr(a, 'get_sources') :
                res.extend(a.get_sources(ctx))
            elif isList(a) :
                for t in a :
                    res.extend(_get_source(self, ctx, t))
            else :
                res.extend(_get_source(self, ctx, a))
    return res

def init(ctx) :
    for m in (font, package, keyboard) :
        if hasattr(m, 'init') :
            m.init(ctx)

@defer
def str_replace(s, i, r):
    return s.replace(i, r)

def onload(ctx) :
    varmap = { 'process' : process, 'create' : create, 'test' : test,
                'cmd' : cmd, 'init' : init, 'str_replace' : str_replace
             }
    for k, v in varmap.items() :
        if hasattr(Context, 'wscript_vars') :
            Context.wscript_vars[k] = v
        else :
            setattr(Context.g_module, k, v)

from . import font, package, keyboard

for m in (font, package, keyboard) :
    if hasattr(m, 'onload') :
        m.onload(Context)
onload(Context)


if __name__ == "__main__":
    main()
